### PR TITLE
MNT: Deprecate changing Figure.number

### LINKED ISSUE #28994: [Bug]: Figure Number Gives Type Error
### Bug summary

It looks like the `Figure` stub does not include the number variable, resulting in a type error.

### Code for reproduction

```Python
import matplotlib.pyplot as plt
screen = plt.figure()
print(screen.number)
```

### Actual outcome

Screenshot

### Expected outcome

Screenshot

### Additional information

I think that this bug is happening because the stub files do not include the number variable for Figure.

### Operating system

Windows 11

### Matplotlib Version

3.9.2

### Matplotlib Backend

tkagg

### Python version

3.12.5

### Jupyter version

_No response_

### Installation

pip

### PR DESCRIPTION
**Pull request by timhoffm**
Historically, pyplot dynamically added a `number` attribute to figure. However, this number must stay in sync with the figure manger. AFAICS overwriting the number attribute does not have the desired effect for pyplot. But there are some repos in GitHub that do change number. So let's take it slow and properly migrate away from writing.

Making the dynamic attribute private and wrapping it in a property allows to maintain current behavior and deprecate write-access.

When the deprecation expires, there's no need for duplicate state anymore and the private _number attribute can be replaced by `return self.canvas.manager.num` if that exists and None otherwise.

Also closes #28994 (dynamic number attribute is not handled by type checkers).

---

@ksunden it would be nice to get this in for 3.10, but if that's too tight, feel free to move to 3.11.

### CODE REVIEW

File: lib/matplotlib/figure.pyi [hunk 1]
```diff
@@ -24,6 +24,8 @@ from matplotlib.lines import Line2D
 from matplotlib.patches import Rectangle, Patch
 from matplotlib.text import Text
 from matplotlib.transforms import Affine2D, Bbox, BboxBase, Transform
+
+from .streamplot import streamplot
```
Comment (tacaswell) [inline lib/matplotlib/figure.pyi:28]: Is this an intended change?

**Comment by jklymak** [comment]
Why are people overwriting the attribute if it "doesn't have the desired effect"?

**Comment by tacaswell** [comment]
> Why are people overwriting the attribute if it "doesn't have the desired effect"?

Either their own reasons or wishful API usage ;)

File: lib/matplotlib/figure.py [hunk 2]
```diff
@@ -2808,6 +2808,33 @@ def axes(self):
 
     get_axes = axes.fget
 
+    @property
+    def number(self):
+        """The figure id, used to identify figures in `.pyplot`."""
+        # Historically, pyplot dynamically added a number attribute to figure.
+        # However, this number must stay in sync with the figure manger.
```
Comment (QuLogic) [inline lib/matplotlib/figure.py:2815]:
```suggestion
        # However, this number must stay in sync with the figure manager.
```

File: lib/matplotlib/figure.py [hunk 3]
```diff
@@ -2808,6 +2808,33 @@ def axes(self):
...
+        # allows to maintain current behavior and deprecate write-access.
+        #
+        # When the deprecation expires, there's no need for duplicate state
+        # anymore and the private _number attribute can be replaced by
+        # `self.canvas.manager.num` if that exists and None otherwise.
+        if hasattr(self, '_number'):
+            return self._number
+        else:
+            raise AttributeError(
+                "'Figure' object has not attribute 'number'. In the future this"
```
Comment (QuLogic) [inline lib/matplotlib/figure.py:2830]:
```suggestion
                "'Figure' object has no attribute 'number'. In the future this"
```

File: lib/matplotlib/figure.py [hunk 4]
```diff
@@ -2808,6 +2808,33 @@ def axes(self):
...
+                "will change to returning 'None' instead.")
+
+    @number.setter
+    def number(self, num):
+        _api.warn_deprecated("3.10", message="Changing 'Figure.number' is deprecated")
```
Comment (QuLogic) [inline lib/matplotlib/figure.py:2835]: `message` needs `%(since)s` and `%(removal)s`.

**Comment by jklymak** [comment]
If number is not to be used externally, why make it public at all?

**Comment by timhoffm** [comment]
Digging into, there are only a few cases that are setting figure numbers:
- `fig.number = `: https://github.com/search?q=%2F%5B%5E%5Cw%5Dfig%5C.number+%3D+%2F+language%3APython+NOT+path%3A%2F%5C.*_pylab_helpers%5C.py%2F+NOT+path%3A%2F%5C.*pyplot%5C.py%2F&type=code
- `figure.number = `: https://github.com/search?q=%2F%5B%5E%5Cw%5Dfigure%5C.number+%3D+%2F+language%3APython+NOT+path%3A%2F%5C.*_pylab_helpers%5C.py%2F+NOT+path%3A%2F%5C.*pyplot%5C.py%2F&type=code

I'm not clear why they do it, but we should not break them without prior warning.

---

> If number is not to be used externally, why make it public at all?

Accessing `number` is not super-useful. Either you work with the number as a standalone label
```
plt.figure(2)
...
plt.close(2)
```
or you keep a reference to the figure, in which case you can use the figure instead of the number:
```
fig = plt.figure(2)
...
plt.close(fig)
```
but since it is used as a core identifier, it feels right to be able to get the information `fig.number` is you really want to.

**Review by tacaswell — APPROVED** [review]

**Review by greglucas — APPROVED** [review]

**Pull request merged**